### PR TITLE
Add `NestingDepthCheck`

### LINKED ISSUE #86: Feature request: add check to have a maximum of 3 nest levels
Could there be a check added to calculate the nest level of a block (being an `if`-statement up to it's own `_endif` or a `_for`-loop)?

```magik
_method a.b
    _if a
    _then
        _if b
        _then
            _if c
            _then
            _endif
        _endif
    _endif
_endmethod
```

```magik
_method a.b
    _if a
    _then
        _if b
        _then
        _endif
    _endif

    _if c
    _then
    _endif
_endmethod
```

This first method has a nesting level of 3, but the second method only has a nesting level of 2 and 1.
With the introduction of this check, we encourage writing cleaner code (and return-early).

### PR DESCRIPTION
**Pull request by sebastiaanspeck**
Resolves #86

### CODE REVIEW

**Comment by sebastiaanspeck** [comment]
Marked as WIP, since I think it can be simplified a little, for the combination of `_for`/`_while` + `_loop`.

Edit: fixed!

File: CHANGES.md [hunk 1]
```diff
@@ -10,6 +10,7 @@
 - Add `magik-typed-lint` to check magik typing from CLI/during CI.
 - Fix not indexing `module.def` files.
 - Rename SwChar16VectorEvaluateInvocationCheck to SwChar16VectorEvaluateInvocationTypedCheck.
+- Add NestingLevelCheck to test if a node does not exceed the maximum nesting level.
```
Comment (StevenLooman) [inline CHANGES.md]: I think you should avoid the term "node" here, but rather a method/procedure/body/....

File: magik-checks/src/main/java/nl/ramsolutions/sw/magik/checks/checks/NestingLevelCheck.java [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+import org.sonar.check.RuleProperty;
+
+/** Check nesting level of nodes. */
+@Rule(key = NestingLevelCheck.CHECK_KEY)
+public class NestingLevelCheck extends MagikCheck {
+
+  @SuppressWarnings("checkstyle:JavadocVariable")
+  public static final String CHECK_KEY = "NestingLevel";
+
+  private static final int DEFAULT_MAXIMUM_NESTING_LEVEL = 3;
```
Comment (StevenLooman) [inline magik-checks/src/main/java/nl/ramsolutions/sw/magik/checks/checks/NestingLevelCheck.java]: I would use the term depth instead of level. Also for the name of the check itself.

File: magik-checks/src/main/java/nl/ramsolutions/sw/magik/checks/checks/NestingLevelCheck.java [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+  protected void walkPreTry(final AstNode node) {
+    this.checkDefinition(node);
+  }
+
+  @Override
+  protected void walkPostTry(final AstNode node) {
+    depthNodes.pop();
+  }
+
+  private void checkDefinition(final AstNode node) {
```
Comment (StevenLooman) [inline magik-checks/src/main/java/nl/ramsolutions/sw/magik/checks/checks/NestingLevelCheck.java]: What (kind of) definition are you checking here? Perhaps you mean something like `checkNestingDepth`?

File: magik-checks/src/main/resources/nl/ramsolutions/sw/sonar/l10n/magik/rules/NestingLevel.html [hunk 4]
```diff
@@ -0,0 +1,34 @@
+<p>The nesting level of nodes should not exceed a defined threshold. Complex code may perform poorly and can be difficult to test thoroughly.</p>
```
Comment (StevenLooman) [inline magik-checks/src/main/resources/nl/ramsolutions/sw/sonar/l10n/magik/rules/NestingLevel.html]: As with the changelog, I would avoid the term "node" as it is unclear for the "regular developer" (if there is any) without any knowledge of ASTs. Also, add that heavily nested code is less comprehensive/difficult to understand.

File: magik-checks/src/main/resources/nl/ramsolutions/sw/sonar/l10n/magik/rules/NestingLevel.json [hunk 5]
```diff
@@ -0,0 +1,18 @@
+{
+  "title": "Nodes should not have too many nesting levels",
```
Comment (StevenLooman) [inline magik-checks/src/main/resources/nl/ramsolutions/sw/sonar/l10n/magik/rules/NestingLevel.json]: Node.

File: magik-checks/src/test/java/nl/ramsolutions/sw/magik/checks/checks/NestingLevelCheckTest.java [hunk 6]
```diff
@@ -0,0 +1,122 @@
...
+import static org.assertj.core.api.Assertions.assertThat;
+
+import java.util.List;
+import nl.ramsolutions.sw.magik.checks.MagikCheck;
+import nl.ramsolutions.sw.magik.checks.MagikIssue;
+import org.junit.jupiter.api.Test;
+
+/** Test NestingLevelCheck. */
+@SuppressWarnings("checkstyle:MagicNumber")
+class NestingLevelCheckTest extends MagikCheckTestBase {
```
Comment (StevenLooman) [inline magik-checks/src/test/java/nl/ramsolutions/sw/magik/checks/checks/NestingLevelCheckTest.java]: What about procedures? Top level constructs in scripts?

Comment (sebastiaanspeck) [reply magik-checks/src/test/java/nl/ramsolutions/sw/magik/checks/checks/NestingLevelCheckTest.java]: > What about procedures? Top level constructs in scripts?

Do you mean to test?:
```
_proc()
    _proc()
        _proc()
        _endproc
    _endproc
_endproc
```

OR 

```
_proc()
    _if a
    _then
        _loop
            _if b
            _then
                _if c
                _then
                _endif
            _endif
        _endloop
    _endif
_endproc
```

Comment (StevenLooman) [reply magik-checks/src/test/java/nl/ramsolutions/sw/magik/checks/checks/NestingLevelCheckTest.java]: The latter.

File: magik-checks/src/main/java/nl/ramsolutions/sw/magik/checks/checks/NestingDepthCheck.java [hunk 7]
```diff
@@ -0,0 +1,69 @@
...
+
+/** Check nesting depth of nodes. */
+@Rule(key = NestingDepthCheck.CHECK_KEY)
+public class NestingDepthCheck extends MagikCheck {
+
+  @SuppressWarnings("checkstyle:JavadocVariable")
+  public static final String CHECK_KEY = "NestingDepth";
+
+  private static final int DEFAULT_MAXIMUM_NESTING_DEPTH = 3;
+  private static final String MESSAGE = "Nesting Depth greater than permitted (%s).";
```
Comment (sebastiaanspeck) [inline magik-checks/src/main/java/nl/ramsolutions/sw/magik/checks/checks/NestingDepthCheck.java]:
```suggestion
  private static final String MESSAGE = "The nesting depth is greater than permitted (%s).";
```

File: magik-checks/src/main/resources/nl/ramsolutions/sw/sonar/l10n/magik/rules/Sonar_way_profile.json [hunk 8]
```diff
@@ -14,6 +14,7 @@
     "LocalImportProcedure",
     "MethodComplexity",
     "MethodLineCount",
+    "NestingDepth",
```
Comment (sebastiaanspeck) [inline magik-checks/src/main/resources/nl/ramsolutions/sw/sonar/l10n/magik/rules/Sonar_way_profile.json:17]: I leave it up to you if we want to enable this check in Sonar by default, since SimplifyIf is also not included in this list.

Comment (StevenLooman) [reply magik-checks/src/main/resources/nl/ramsolutions/sw/sonar/l10n/magik/rules/Sonar_way_profile.json:17]: Yes, lets add it.

File: magik-checks/src/main/java/nl/ramsolutions/sw/magik/checks/checks/NestingDepthCheck.java [hunk 9]
```diff
@@ -0,0 +1,69 @@
...
+  /** Maximum nesting depth of node. */
+  @RuleProperty(
+      key = "maximum nesting depth",
+      defaultValue = "" + DEFAULT_MAXIMUM_NESTING_DEPTH,
+      description = "Maximum nesting depth",
+      type = "INTEGER")
+  @SuppressWarnings("checkstyle:VisibilityModifier")
+  public int maximumNestingDepth = DEFAULT_MAXIMUM_NESTING_DEPTH;
+
+  private Deque<AstNode> depthNodes = new ArrayDeque<>();
```
Comment (StevenLooman) [inline magik-checks/src/main/java/nl/ramsolutions/sw/magik/checks/checks/NestingDepthCheck.java]: This can be final.

File: magik-checks/src/main/java/nl/ramsolutions/sw/magik/checks/checks/NestingDepthCheck.java [hunk 10]
```diff
@@ -0,0 +1,69 @@
...
+  }
+
+  private void checkNestingDepth(final AstNode node) {
+    depthNodes.push(node);
+
+    if (depthNodes.size() == DEFAULT_MAXIMUM_NESTING_DEPTH + 1) {
+      AstNode lastNode = depthNodes.peek();
+      this.addIssue(lastNode, String.format(MESSAGE, DEFAULT_MAXIMUM_NESTING_DEPTH));
+    }
+  }
```
Comment (StevenLooman) [inline magik-checks/src/main/java/nl/ramsolutions/sw/magik/checks/checks/NestingDepthCheck.java:69]: Do we want to add more constructs to the check, like `_block`, `_protect`, `_lock`, `_catch`, or embedded procedures?

File: magik-checks/src/main/resources/nl/ramsolutions/sw/sonar/l10n/magik/rules/NestingDepth.html [hunk 11]
```diff
@@ -0,0 +1,34 @@
...
+_endif
+</pre>
+<h2>Compliant Solution</h2>
+<pre>
+_if a _is _unset _then _return b _endif
+
+_for x _over a.fast_elements()
+_loop
+  _if x _isnt _unset
+  _then
```
Comment (StevenLooman) [inline magik-checks/src/main/resources/nl/ramsolutions/sw/sonar/l10n/magik/rules/NestingDepth.html:27]: You could further improve (subjective, of course) that you can rewrite this as:
```
_for x _over a.fast_elements()
_loop
  _if x _is _unset
  _then
    _continue
  _endif

  _for child _over x.fast_elements()
  _loop
    _return child
  _endloop
_endloop
```

Or even write the `_if x ...` on a single line. Though for clarity I wouldn't do this.

File: magik-checks/src/main/java/nl/ramsolutions/sw/magik/checks/checks/NestingDepthCheck.java (hunk 10 above)
Comment (sebastiaanspeck) [reply magik-checks/src/main/java/nl/ramsolutions/sw/magik/checks/checks/NestingDepthCheck.java:69]: I was thinking about that as well, but then having the same `_if` in a `_block` has an extra NestingDepth compared to having it in a `_method`.

File: magik-checks/src/main/java/nl/ramsolutions/sw/magik/checks/checks/NestingDepthCheck.java (hunk 9 above)
Comment (sebastiaanspeck) [inline magik-checks/src/main/java/nl/ramsolutions/sw/magik/checks/checks/NestingDepthCheck.java]:
```suggestion
  private final Deque<AstNode> depthNodes = new ArrayDeque<>();
```

File: magik-checks/src/main/resources/nl/ramsolutions/sw/sonar/l10n/magik/rules/NestingDepth.html [hunk 12]
```diff
@@ -0,0 +1,34 @@
...
+    _for child _over x.fast_elements()
+    _loop
+      _return child
+    _endloop
+  _endif
```
Comment (sebastiaanspeck) [inline magik-checks/src/main/resources/nl/ramsolutions/sw/sonar/l10n/magik/rules/NestingDepth.html]:
```suggestion
  _if x _is _unset
  _then
    _continue
  _endif

  _for child _over x.fast_elements()
  _loop
    _return child
  _endloop
```

File: magik-checks/src/main/java/nl/ramsolutions/sw/magik/checks/checks/NestingDepthCheck.java (hunk 10 above)
Comment (StevenLooman) [reply magik-checks/src/main/java/nl/ramsolutions/sw/magik/checks/checks/NestingDepthCheck.java:69]:
```magik
_method a.b
	...

	_block
		_if a
		_then
			...
		_endif
	_endblock

	...
_endmethod
```

I think this should be marked, as the depth is increased by the block.

Comment (sebastiaanspeck) [reply magik-checks/src/main/java/nl/ramsolutions/sw/magik/checks/checks/NestingDepthCheck.java:69]: Yes, but in the situation:

```
_block
	_if a
	_then
		...
	_endif
_endblock
```

`_block` shouldn't be counted IMO.

Comment (StevenLooman) [reply magik-checks/src/main/java/nl/ramsolutions/sw/magik/checks/checks/NestingDepthCheck.java:69]: And in case of a method, should the top body be counted?

Comment (sebastiaanspeck) [reply magik-checks/src/main/java/nl/ramsolutions/sw/magik/checks/checks/NestingDepthCheck.java:69]: No. So we need to keep track of the startNode.

Comment (StevenLooman) [reply magik-checks/src/main/java/nl/ramsolutions/sw/magik/checks/checks/NestingDepthCheck.java:69]: I believe you are saying the top should never be counted, whether it is a method or a block. Which is also what I am saying.

Any deeper block (or rather, body) should be counted, I think.

Comment (StevenLooman) [reply magik-checks/src/main/java/nl/ramsolutions/sw/magik/checks/checks/NestingDepthCheck.java:69]: I think the easiest is to (only) use the `walkPreBody()` + `walkPostBody()` methods, incrementing and decrementing currentDepth respectively. This way there is no need to keep a stack. But you do keep track of all constructs which increment the nesting level.

Then, ignore the top body. I.e., a method which returns the current depth, something such as:
```java
int getCurrentDepth() {
  return this.currentDepth - 1;
}
```

Comment (sebastiaanspeck) [reply magik-checks/src/main/java/nl/ramsolutions/sw/magik/checks/checks/NestingDepthCheck.java:69]: Thanks for the hint. Will implement soon.

Comment (sebastiaanspeck) [reply magik-checks/src/main/java/nl/ramsolutions/sw/magik/checks/checks/NestingDepthCheck.java:69]: I tried it, but no matter how, it fails one or more tests.

Comment (StevenLooman) [reply magik-checks/src/main/java/nl/ramsolutions/sw/magik/checks/checks/NestingDepthCheck.java:69]: I see, test `testNotExceedingMaximumNestingDepthWithEarlyReturns` fails. This is because the `_if b _then _leave _endif` is also adding a level. Do you expect that this does not add to the nesting depth?

I also see that the check tests for `node.getToken() != null`. If the body is empty, this body is ignored. Following this, several tests have if constructs with empty bodies. Now the nesting depth is already too deep at the lowest level, but perhaps it is better to add a statement to the body.

Comment (StevenLooman) [reply magik-checks/src/main/java/nl/ramsolutions/sw/magik/checks/checks/NestingDepthCheck.java:69]: I've added some changes. The question about the failing test still stands!

File: magik-checks/src/main/resources/nl/ramsolutions/sw/sonar/l10n/magik/rules/Sonar_way_profile.json (hunk 8 above)
Comment (sebastiaanspeck) [reply magik-checks/src/main/resources/nl/ramsolutions/sw/sonar/l10n/magik/rules/Sonar_way_profile.json:17]: So add SimplifyIf as well? (In a separate PR)

Comment (StevenLooman) [reply magik-checks/src/main/resources/nl/ramsolutions/sw/sonar/l10n/magik/rules/Sonar_way_profile.json:17]: I have added it.

File: magik-checks/src/main/java/nl/ramsolutions/sw/magik/checks/checks/NestingDepthCheck.java (hunk 10 above)
Comment (sebastiaanspeck) [reply magik-checks/src/main/java/nl/ramsolutions/sw/magik/checks/checks/NestingDepthCheck.java:69]: IMO `_if b _then _leave _endif` is an early-return and would not add to the nesting depth. Otherwise, how can you avoid the nesting depth nightmare?

Comment (StevenLooman) [reply magik-checks/src/main/java/nl/ramsolutions/sw/magik/checks/checks/NestingDepthCheck.java:69]: There are (at least) two ways to look at this:
1. It should add to the nesting depth. If you're at that depth, it is either time to refactor the method, or to increase the maximum depth. This is the check in its current (and purest) form.
2. Allow a body with only a `_leave`/`_continue`/`_return` (others?) for the early-return cases, essentially adding an exception.

I think both should be supported, but configurable. Can you add a boolean parameter to the check with which you can configure this?

Comment (sebastiaanspeck) [reply magik-checks/src/main/java/nl/ramsolutions/sw/magik/checks/checks/NestingDepthCheck.java:69]: Done!

**Comment by StevenLooman** [comment]
LGTM. Thanks!
